File: lab/repro/bun-ffi-stack-args/repro.ts
// Bun 1.3.14 macOS arm64: bun:ffi writes each STACK argument to its own
// 8-byte slot; the Apple arm64 ABI packs sub-8-byte stack args at natural
// size. Any signature whose 9th+ int-class args include a sub-8-byte type
// followed by more args gets shifted garbage in the callee.
//
// Build: cc -dynamiclib -o repro.dylib repro.c
// Run:   bun repro.ts
//
// conv2d_echo(int64* res, u64, u64, int, int, int, int, int, int, int, u64)
// Registers hold res..dilation_0 (8 args). Stack (Apple natural packing):
//   dilation_1 @ sp+0 (4B) | groups @ sp+4 (4B) | stream @ sp+8 (8B)
// Bun writes: dilation_1 @ sp+0, groups @ sp+8, stream @ sp+16
// → callee reads groups=0 (high half of slot 0), stream=<groups value>.
import { dlopen, ptr } from "bun:ffi";

const { symbols: C } = dlopen(new URL("repro.dylib", import.meta.url).pathname, {
  conv2d_echo: {
    args: ["ptr","u64","u64","i32","i32","i32","i32","i32","i32","i32","u64"],
    returns: "i32",
  },
});
const res = new BigInt64Array(10);
C.conv2d_echo(ptr(res), 111n, 222n, 2, 3, 4, 5, 6, 7, 8, 999n);
const got = Array.from(res).map(Number);
const want = [111, 222, 2, 3, 4, 5, 6, 7, 8, 999];
console.log("got :", got.join(","));
console.log("want:", want.join(","));
console.log(got.join(",") === want.join(",") ? "OK" : "CORRUPTED (bug present)");

File: lab/repro/bun-ffi-stack-args/repro.c
#include <stdint.h>
// exact mlx_conv2d shape: (ptr, u64, u64, int×7, u64) -> int
int conv2d_echo(int64_t* res, uint64_t a, uint64_t b, int s0, int s1,
                int p0, int p1, int d0, int d1, int g, uint64_t st) {
  res[0]=(int64_t)a; res[1]=(int64_t)b; res[2]=s0; res[3]=s1; res[4]=p0;
  res[5]=p1; res[6]=d0; res[7]=d1; res[8]=g; res[9]=(int64_t)st;
  return 42;
}

// exact mlx_conv3d shape: (ptr, u64, u64, int×10, u64) -> int
int conv3d_echo(int64_t* res, uint64_t a, uint64_t b,
                int s0, int s1, int s2, int p0, int p1, int p2,
                int d0, int d1, int d2, int g, uint64_t st) {
  res[0]=(int64_t)a; res[1]=(int64_t)b; res[2]=s0; res[3]=s1; res[4]=s2;
  res[5]=p0; res[6]=p1; res[7]=p2; res[8]=d0; res[9]=d1; res[10]=d2;
  res[11]=g; res[12]=(int64_t)st;
  return 43;
}
